chain/network/src/concurrency/ctx/mod.rs

```diff
@@ -0,0 +1,220 @@
+/// eventually without requiring any wake()).
+///
+/// Ctx is similar to a rust lifetime but for runtime:
+/// Ctx is expected to be passed down the call stack and to the spawned tokio subtasks.
+/// At any level of the call stack the Ctx can be narrowed down via `Ctx::with_cancel` (which
+/// allows the routine to cancel context of subroutines it spawns) or via
+/// `Ctx::with_timeout`/`Ctx::with_deadline`, which cancels the scope automatically after a given time.
+/// If is NOT possible to extend the context provided by the caller - a subtask is expected
+/// to adhere to the lifetime of its context and finish as soon as it gets canceled (or earlier).
+#[derive(Clone)]
+pub(super) struct Ctx(Arc<Inner>);
+
+impl Ctx {
+    /// Constructs a new context:
+    /// * without a parent
+    /// * with real clock
+    /// * with infinite deadline
+    ///
+    /// It should be called directly from main.rs.
+    pub(crate) fn new(clock: crate::time::Clock) -> Ctx {
+        return Ctx(Arc::new(Inner {
+            canceled: signal::Once::new(),
+            deadline: crate::time::Deadline::Infinite,
+            clock,
+            _parent: None,
+        }));
+    }
+
+    pub fn cancel(&self) {
+        self.0.canceled.send();
+    }
+
+    pub fn sub(&self, deadline: crate::time::Deadline) -> Ctx {
+        let child = Ctx(Arc::new(Inner {
+            canceled: signal::Once::new(),
+            clock: self.0.clock.clone(),
+            deadline: std::cmp::min(self.0.deadline, deadline),
+            _parent: Some(self.0.clone()),
+        }));
+        tokio::spawn({
+            let clock = self.0.clock.clone();
+            let deadline = child.0.deadline;
+            let parent = self.0.canceled.clone();
+            let child = child.0.canceled.clone();
+            async move {
+                tokio::select! {
+                    _ = clock.sleep_until_deadline(deadline) => child.send(),
+                    _ = parent.recv() => child.send(),
+                    _ = child.recv() => {}
+                }
+            }
+        });
+        child
+    }
+}
+
+/// Awaits for the current task to get canceled.
+pub async fn canceled() {
+    local().0.canceled.recv().await
+}
+
+/// Check if the context has been canceled.
+pub fn is_canceled() -> bool {
+    local().0.canceled.try_recv()
+}
+
+/// The time at which the local context will be canceled.
+/// The current task should use it to schedule its work accordingly.
+/// Remember that this is just a hint, because the local context
+/// may get canceled earlier.
+pub fn get_deadline() -> crate::time::Deadline {
+    local().0.deadline
+}
+
+pub(super) fn local() -> Ctx {
+    CTX.with(|ctx| unsafe { &*ctx.get() }.clone())
+}
+
+impl Ctx {
+    /// Awaits until f completes, or the context gets canceled.
+    /// f is required to be cancellable.
+    async fn wait<F: Future>(&self, f: F) -> OrCanceled<F::Output> {
+        tokio::select! {
```

Comment:
CtxFuture sets the local context for the given future. ctx::wait wraps a future into tokio::select to include the possibility of the task-local context being canceled. These are 2 complementary but different features.

Reply:
Oh I missed the fact that this is not `pub`.

Then maybe it’d be better to just inline it into its only caller? I was going from the fact that `ctx.wait(f)` waiting based on the task-local context instead of `ctx` was surprising, but it’ll just never be called this way.